### ISSUE TITLE
Incremental build is not working properly for projects have large dependencies.

### ISSUE THREAD

**Issue by pocorall** [pfn/android-sdk-plugin#200]
For my project, It has been long time since incremental build is broken.

After I successfully clean build the project, change the source code (do the trivial thing like changing some character in string, to avoid possible complicated proguard problem) and re-build fails with this message:

```
(junit.swingui.TestSelector$4) that doesn't come with an
associated EnclosingMethod attribute. This class was probably produced by a
compiler that did not target the modern .class file format. The recommended
solution is to recompile the class from source, using an up-to-date compiler
and without specifying any "-target" type options. The consequence of ignoring
this warning is that reflective operations on this class will incorrectly
indicate that it is *not* an inner class.
```

After for a while, the dex process is dead with `OutOfMemoryError`:

```
java.lang.OutOfMemoryError: GC overhead limit exceeded
        at com.android.dx.dex.code.BlockAddresses.setupArrays(BlockAddresses.jav
a:139)
        at com.android.dx.dex.code.BlockAddresses.<init>(BlockAddresses.java:57)

        at com.android.dx.dex.code.RopTranslator.<init>(RopTranslator.java:125)
        at com.android.dx.dex.code.RopTranslator.translate(RopTranslator.java:10
3)
        at com.android.dx.dex.cf.CfTranslator.processMethods(CfTranslator.java:3
17)
        at com.android.dx.dex.cf.CfTranslator.translate0(CfTranslator.java:137)
        at com.android.dx.dex.cf.CfTranslator.translate(CfTranslator.java:93)
        at com.android.dx.command.dexer.Main.processClass(Main.java:729)
        at com.android.dx.command.dexer.Main.processFileBytes(Main.java:673)
        at com.android.dx.command.dexer.Main.access$300(Main.java:83)
```

Unfortunately, I don't know how to reproduce it with a simple example project, but I found what causes the problem. The problem was started to appear from `0.3.21`, so I revert the change, then it works OK. I propose a pull request about this.

I tried to understand the change made in `0.3.21`, but I have very limited knowledge about the internal process of this plugin. Please let me understand the implication of the change of that version.

**Comment by pocorall** [comment]
I couldn't reproduce the same error with example project, but I found that a notable thing. What I did is:

* Clone https://github.com/pocorall/hello-scaloid-sbt.git
* `sbt run`:
```
> run
...
[info] Generating dex, incremental=false, multidex=false
[info] dex method count: 4737
[info] Packaged: hello-scaloid-sbt-debug-unaligned.apk (189.73KB)
[info] Debug package does not need signing: hello-scaloid-sbt-debug-unaligned.apk
[info] zipaligned: hello-scaloid-sbt-debug.apk
[info] Installing to LG-V500 (LGV5005047858a)...
[info] [hello-scaloid-sbt-debug.apk] Install finished: 189.73KB in 3.54s. 53.61KB/s
[info] Running on LG-V500 (LGV5005047858a)...
[info] Starting: Intent { cmp=scaloid.example/.HelloScaloid }
[success] Total time: 19 s, completed 2015. 8. 24 오후 7:11:14
```

* `sbt run` again:

```
> run
[info] Processing resources
[info] [debug] cache hit, skipping proguard!
[info] Generating dex, incremental=false, multidex=false
[info] dex method count: 4955
[info] Packaged: hello-scaloid-sbt-debug-unaligned.apk (199.98KB)
[info] Debug package does not need signing: hello-scaloid-sbt-debug-unaligned.apk
[info] zipaligned: hello-scaloid-sbt-debug.apk
[info] Installing to LG-V500 (LGV5005047858a)...
[info] [hello-scaloid-sbt-debug.apk] Install finished: 199.98KB in 2.55s. 78.45KB/s
[info] Running on LG-V500 (LGV5005047858a)...
[info] Starting: Intent { cmp=scaloid.example/.HelloScaloid }
[success] Total time: 5 s, completed 2015. 8. 24 오후 7:11:20
```

The method count is different. When I do this again with a modified plugin being applied the pull request #201, the method count is the same.

**Comment by pfn** [comment]
Turning off incremental dex for the initial proguard cache run breaks proguard cache.

The proper fix for what you are encountering is to increase `dexMaxHeap in Android` and then add the offending packages to `proguardCache in Android` once you are able to get past dex OOME, because you are definitely over dex limit during proguard cache. Alternatively, disable proguard cache.

**Comment by pocorall** [comment]
Increasing dexMaxHeap removes OOME, thanks. And, as you said, it fails because of exceeding dex method limit:

```
trouble writing output: Too many method references: 372248; max is 65536.
You may try using --multi-dex option.
References by package:
     6 android.accessibilityservice
     5 android.accounts
    10 android.animation
   293 android.app
     3 android.appwidget
   208 android.content
    29 android.content.pm
    53 android.content.res
    53 android.database
    19 android.database.sqlite
    32 android.gesture
   189 android.graphics
    58 android.graphics.drawable
     2 android.graphics.drawable.shapes
     3 android.graphics.pdf
...	 
```

The method count of clean rebuild of our project is around 40,000, but this case exceeds 9 times more than the clean rebuild. I have some more questions:

* Why does it exceeds so much?
* Pull request #201 have no problem with this method count. Why?
* You said that the version prior to 1.3.21 breaks proguard cache during the initial proguard cache (hit) run. What exactly it is? Is that spoil `classes.proguard.jar` or `proguard-cache-xxxxxx.jar`? Neither things seems be changed for me.
* Finally, a suggestion: Writing an introductory article about internal process would be very appreciated.

**Comment by pfn** [comment]
The pr completely bypasses dex. That's why it builds with no change in
method count.

Incremental dex fails in the presence of proguard cache and behaves as if
dex does not need to run.

--

**Comment by pocorall** [comment]
Your answer looks like a riddle for me :(

I still don't understand why the method count is increased 9 times more at the second run, with current version of this plugin.

**Comment by pfn** [comment]
Everything is passing through without getting proguarded. Only scala.* is
cached by default.

--

**Comment by pocorall** [comment]
Yes I was turned off the proguardCache settings for the last experiment. Now I applied the setting below:

```
proguardCache in Android ++= Seq("org.scaloid", "com.google", "com.facebook", "com.googlecode", "com.coremedia", "org.achartengine", "junit")
```

This removes `that doesn't come with an
associated EnclosingMethod attribute` error, in which proguard caching "junit" seems affected that. And the method count of the second run is now 272,420. It is 6 times of the first run.

The size of `proguard-cache-xxxxx.jar` and `classes.proguard.jar` is around 4MB and 7MB respectively. It exceeds the size of `.dex`. What exactly does/should it contains? how different these two jars? and why does it is significant larger than `bin/classes.jar` at the clean build (about 2MB)?

**Comment by pfn** [comment]
How proguard cache works:

Run proguard, take everything that prefix matches the rules in
proguardCache and store them in proguard-cache.jar

Everything else is not minimized by proguard.

On a cache hit, take proguard-cache.jar and the remaining jars and pass
directly to dex.

Therefore, as a result, dex method counts for a proguard cache hit must
always be higher than a full proguard run.

To fix running out of methods in a cache hit, take the highest offending
packages and add them to proguardCache.

--

**Comment by pocorall** [comment]
Whoaa! Finally I got a right answer for my problem. Thanks!

**Everything else is not minimized by proguard** was the key. I really had no idea about this. After adding most packages in proguardCache, finally it works fine. Instead I got a long list for `proguardCache in Android` setting.

How about making some exclusion filter for proguardCache? A description like "Include all except `com.soundcorset` package" will make maintaining this list much easier.
